core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -339,7 +341,7 @@ private[spark] class TaskSchedulerImpl(
           // We lost this entire executor, so remember that it's gone
           val execId = taskIdToExecutorId(tid)
 
-          if (executorIdToTaskCount.contains(execId)) {
+          if (executorIdToRunningTaskIds.contains(execId)) {
             reason = Some(
               SlaveLost(s"Task $tid was lost, so marking the executor as lost as well."))
             removeExecutor(execId, reason.get)
```

Comment:
More archaeology:

It looks like `TaskState.LOST` was introduced in Spark 0.6.0 as part of a refactoring to make the cluster scheduler pluggable: https://github.com/apache/spark/commit/e72afdb817bcc8388aeb8b8d31628fd5fd67acf1. That commit is from July 2012.

At the time, standalone mode didn't even exist and the schedulers were Mesos and local mode, and only Mesos fine-grained mode was supported. The only way to get a `TaskState.LOST` state was to convert the Mesos task loss state to it.